src/mscorlib/src/System/Buffer.cs

```diff
@@ -256,326 +260,146 @@ internal unsafe static void Memcpy(byte* dest, byte* src, int len)
         // This method has different signature for x64 and other platforms and is done for performance reasons.
         internal unsafe static void Memmove(byte* dest, byte* src, nuint len)
         {
-            // P/Invoke into the native version when the buffers are overlapping and the copy needs to be performed backwards
-            // This check can produce false positives for lengths greater than Int32.MaxInt. It is fine because we want to use PInvoke path for the large lengths anyway.
-
-            if ((nuint)dest - (nuint)src < len) goto PInvoke;
-
-            // This is portable version of memcpy. It mirrors what the hand optimized assembly versions of memcpy typically do.
-            //
-            // Ideally, we would just use the cpblk IL instruction here. Unfortunately, cpblk IL instruction is not as efficient as
-            // possible yet and so we have this implementation here for now.
-
-            // Note: It's important that this switch handles lengths at least up to 22.
-            // See notes below near the main loop for why.
-
-            // The switch will be very fast since it can be implemented using a jump
-            // table in assembly. See http://stackoverflow.com/a/449297/4077294 for more info.
+            byte* srcEnd = src + len;
+            byte* destEnd = dest + len;
+            
+#if AMD64 && !PLATFORM_UNIX 
+            const nuint CopyThreshold = 1024;
+#else
+            const nuint CopyThreshold = 2048;
+#endif // AMD64 && !PLATFORM_UNIX
 
-            switch (len)
+            // PInvoke to the native version when the buffers are overlapping.            
+            if ((((nuint)dest <= (nuint)src) && ((nuint)destEnd > (nuint)src)) || 
```

Comment:
```src + len``` needs to be re-computed in quite a few places and so created ```srcEnd``` to avoid this. Had optimized the original condition to ```if ((nuint)dest < (nuint)srcEnd)``` and used it in the PR submitted last week. Revised it in favor of a more thorough overlap check as suggested in one of your CR comments.